storage/src/desktop/storage_reference_desktop.cc

```diff
@@ -279,18 +277,95 @@ Future<size_t> StorageReferenceInternal::GetBytes(void* buffer,
+    StorageReferenceFn internal_function_reference,
+    SendRequestFunct send_request_funct, SafeFutureHandle<size_t> final_handle,
+    double max_retry_time_seconds) {
+  BlockingResponse* first_response = send_request_funct();
+  std::thread async_retry_thread(
+      &StorageReferenceInternal::AsyncSendRequestWithRetry, this,
+      internal_function_reference, send_request_funct, final_handle,
+      first_response, max_retry_time_seconds);
+  async_retry_thread.detach();
+}
+
+const int kInitialSleepTimeMillis = 1000;
+const int kMaxSleepTimeMillis = 30000;
+
+// In a separate thread, repeatedly send Rest requests until one succeeds or a
+// maximum amount of time has passed.
+void StorageReferenceInternal::AsyncSendRequestWithRetry(
+    StorageReferenceFn internal_function_reference,
+    SendRequestFunct send_request_funct, SafeFutureHandle<size_t> final_handle,
+    BlockingResponse* response, double max_retry_time_seconds) {
+  auto* future_api = future();
+  firebase::FutureBase internal_future;
+  auto end_time = std::chrono::steady_clock::now() +
+                  std::chrono::duration<double>(max_retry_time_seconds);
+  auto current_sleep_time = std::chrono::milliseconds(kInitialSleepTimeMillis);
+  while (true) {
+    internal_future = future_api->LastResult(internal_function_reference);
+    // Wait for completion, then check status and error.
+    while (internal_future.status() == firebase::kFutureStatusPending) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+    // For any request that succeeds or fails in a non-retryable way, don't
+    // bother retrying.
+    if (internal_future.status() != firebase::kFutureStatusComplete ||
+        !IsRetryableFailure(response->status())) {
+      break;
+    }
+    // Interrupt the loop if the retry deadline has been reached
+    auto current_time = std::chrono::steady_clock::now();
+    if (current_time + current_sleep_time > end_time) {
+      break;
+    }
+    // Sleep for an exponentially increasing duration, then retry the request.
+    std::this_thread::sleep_for(current_sleep_time);
+    if (current_sleep_time < std::chrono::milliseconds(kMaxSleepTimeMillis)) {
+      current_sleep_time *= 2;
+    }
+    response = send_request_funct();
+  }
+  // Copy from the internal future to the final future.
+  Future<size_t> buffer_index_future =
+      static_cast<const Future<size_t>&>(internal_future);
+  if (buffer_index_future.result() != nullptr) {
+    future_api->CompleteWithResult(final_handle, internal_future.error(),
+                                   *(buffer_index_future.result()));
+  } else {
+    future_api->Complete(final_handle, internal_future.error(),
+                         internal_future.error_message());
+  }
+}
 
-  return GetBytesLastResult();
+// Returns whether or not an http status represents a failure that should be retried.
+bool StorageReferenceInternal::IsRetryableFailure(int httpStatus) {
```

Comment:
Here is one ugly way to set up the testing override:

You could set up a global variable here:
static bool g_retry_all_futures_for_testing = false;

And then in the integration_test.cc:
#if FIREBASE_PLATFORM_DESKTOP
namespace firebase {
namespace storage {
namespace internal {
extern bool g_retry_all_futures_for_testing;
}
}
}
#endif  // FIREBASE_PLATFORM_DESKTOP


And in the test that forces retrying futures, inside an #if FIREBASE_PLATFORM_DESKTOP:
bool old_value = g_retry_all_futures_for_testing;
g_retry_all_futures_for_testing = true;
// ..... do all the testing here  .....
g_retry_all_futures_for_testing = old_value;

You could also make the thing you extern a function rather than a global.

Reply:
(also, feel free to name it better, like retry_all_http_statuses or whatever)

Reply:
Done. 

I made the option only retry 404 failures since I don't want it to retry successful statuses like 200 or 204 and so retrying on a single failure status is the simplest solution.

I am not quite sure if there is a best place in each of these files to place the code snippets that you shared. I put the g_retry_file_not_found__for_testing definition right about the IsRetryable method, and the extern line near the top of the integration test.

I wrapped the entire test in an IF FIREBASE_PLATFORM_DESKTOP check.

Reply:
This part looks good to me.